target/linux/mvebu/files-4.14/arch/arm/boot/dts/armada-385-wd-mcex2u.dts

```diff
@@ -0,0 +1,419 @@
+/*
+ * Device Tree file for Marvell Armada 385 evaluation board
+ * (DB-88F6820)
+ *
+ *  Copyright (C) 2013 Marvell
+ *
+ * Gregory CLEMENT <gregory.clement@free-electrons.com>
+ * Thomas Petazzoni <thomas.petazzoni@free-electrons.com>
+ *
+ * This file is licensed under the terms of the GNU General Public
+ * License version 2.  This program is licensed "as is" without any
+ * warranty of any kind, whether express or implied.
+ */
```

Comment:
Instead of this text, add SPDX license identifier at the beginning of the file.

Reply:
Loading file '/mac_addr' to addr 0x02000000 with size 18 (0x00000012)...
They load it into memory and then clearly use it for initializing eth because 'dhcp' in u-boot sends correct MAC. But I can't see it anywhere in 'env print -a'. And then they use mounted config partition later in the boot process for getting the address.

Reply:
Can You check if adding
````
aliases {
        ethernet0 = &eth0;
};
````
node (below `chosen`) will work? We do this for ESPRESSObin boards and it uses same network driver, so theoretically should work.

Reply:
Nope, no cigar. Might it require some property override, like memory address?

Reply:
One thing has changed though:
[    0.953300] mvneta f1034000.ethernet eth0: Using random mac address 96:88:08:66:ae:5d
Became
[    1.007963] mvneta f1034000.ethernet eth0: Using hardware mac address 00:50:43:f6:1a:24
But the MAC is wrong

Reply:
Can You try changing `ethernet0` to `ethernet1` and after `ethernet2`?

Reply:
Nope, ethernet1 and ethernet2 changes nothing. Can you elaborate what are we looking here for? Where should U-Boot place the MAC? Maybe I could just write a bootscript for this.

For some reason I can't attach environment log here. I'll try to put it elsewhere.

Reply:
The `ethernet0 = &eth0` tells the U-Boot to replace MAC address of eth0 with data from ethernet0. In particular when ethernet0 is specified it takes MAC from ethaddr or eth0addr variable, when ethernet1 it should take address from eth1addr, when ethernet2 from eth2addr and so on. I thought that maybe U-Boot stored the proper MAC in ethXaddr value, so that's why i asked about next tests. You can try it out, set dts back to "ethernet0 = &eth0" and change the value of ethaddr to Your MAC, save Your environment and reboot. It might not work because this MAC:
`mvneta f1034000.ethernet eth0: Using hardware mac address 00:50:43:f6:1a:24`
isn't the same as value of ethaddr in environment You posted.
As I look more on the issue, it seems like Your approach with setting MAC is a proper one.

Reply:
So, saving real MAC to `eth2addr` in U-Boot does the trick. But this U-Boot version lacks `setexpr` and I have no other idea about how to write MAC, which is loaded into memory from UBI partition, into the environment variable. I guess bootscript it's really a dead end.

Reply:
can this uboot save the environment vars with **saveenv**? 

I see the "       _U-Boot Environment:      0x00000000:0x00080000 (NAND)_" in the boot logs so I think there is a writable environment.

Also check if you have "*setenv*", on Kirkwood devices (Marvell hardware) the uboot has "*setenv*" command to set env variables.

Also if you can't write it from uboot you can write it from OpenWrt with the tool in the uboot-envtools package (if you can't automate it you can ask the user to do this as part of the installation procedure)

Reply:
There is no *saveenv*. While *setenv* is present, and does the trick when I manually save right MAC into the variable (I guess saveenv is just not needed). But we are looking for an automated (and more elegant than mounting config partition in preinit) way to do this.

Reply:
> So, saving real MAC to eth2addr in U-Boot does the trick.

That's great, with that we can do something.

> But this U-Boot version lacks setexpr and I have no other idea about how to write MAC, which is loaded into memory from UBI partition, into the environment variable.

We can use for that uboot-envtools, instead of fiddling with U-Boot command line and bootscripts. The idea would be, add alias node to dts; on first boot, in preinit script, read out MAC from config partition; set the eth2addr variable and additional one like for example mac_set=1.  The additional variable would be for testing if the U-Boot environment was reset, if yes script will redo whole procedure. That way we won't mount config partition on every boot, it should speed up the boot process a bit.

BTW, in `u-boot-2013.01-2014_T3.0p2/board/mv_ebu/a38x/mv_main_a38x.c` from WDMyCloud_Ex2Ultra_GPL_v2.31.163_20181219 tarball is where the "magic" happens regarding MAC address, that is also a reason why I asked for U-Boot environment to check if anything is writing to `0x02000000` and as it stands, yes it does, so we couldn't read it from there.

Reply:
OK. I surrender.

First of all, after a power cycle manually set `eth2addr` disappears, which is expectable as I did not save it with `saveenv` (because it is absent). Then I tried to locate where these 0x80000 of env are actually stored. The answer is: *nowhere*.

They define it's offset as `#define CONFIG_ENV_OFFSET nand_get_env_offs()`, which boils down to `offset = CONFIG_UBOOT_SIZE + CONFIG_SPARE_AREA` or somewhere further away if `CONFIG_SKIP_BAD_BLOCK` is defined.

Given `#define CONFIG_UBOOT_SIZE   0x100000` and `#define CONFIG_SPARE_AREA   0x400000`, we have offset no less than 0x500000. But the whole size of the U-Boot partition is 0x500000, next, without any gaps, goes kernel. Huh?

Reply:
Screw that. They actually define is as 0 somewhere, but it does not matter. They use environment as local variables, not storing them anywhere and not reading them from anywhere, as far as I can see from their U-Boot sources.

Reply:
> Screw that. They actually define it as 0 somewhere, but it does not matter. They use environment as local variables, not storing them anywhere and not reading them from anywhere, as far as I can see from their U-Boot sources.

True, that was my derp moment, sorry for pushing You down the rabbit hole. Indeed the U-Boot recreates the environment (with values from the file I mentioned) on each reboot and doesn't store it anywhere. It's pity they decided not to rely on U-Boot environment for storing configuration.